### PR TITLE
[11.0] [MIG] purchase_order_variant_mgmt

### PR DESCRIPTION
**Pull request by javierjcf**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please squash together adjacent "OCA Transbot.." commits using https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

**Comment by javierjcf** [comment]
It's squashed now.

**Comment by pedrobaeza** [comment]
There is another migration in progress in https://github.com/OCA/purchase-workflow/pull/594, but this is the right repository.

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant_view.xml
Comment (grindtildeath) [inline purchase_order_variant_mgmt/wizard/purchase_manage_variant_view.xml]: Please fix the space here.

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (grindtildeath) [inline purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: Is the `browse` here really needed ? IMO would look cleaner to use `self.env['purchase.order.line']`

File: purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py
Comment (grindtildeath) [inline purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: Was this change really needed ? IMO looks safer to assert a quantity according to its product than its order in a recordset.

**Review by grindtildeath — COMMENTED** [review]
Thanks for the port, I tested and works well. Just a few minor things here and can you please split the readme into parts ?

Then, this is not really to do here in the migration, but since it appears here I state it : it would be great to change the message "Sorry no matrix data to display." to something more specific to the context like "Select a product first" here. However it seems there's no hook to define a specific message in the matrix widget module, but having the possibility to set the message in the XML field definition would be IMO a very nice improvement to the widget. cc @simahawk @arkostyuk

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant_view.xml
Comment (javierjcf) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant_view.xml]: fixed

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (javierjcf) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: I did the change because in the sale_order_variant_mgmt (already migrated) appears that way. I don't know the reason and i copy it. It is with the self.env['purchase.order.line'] now.

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant_view.xml
Comment (javierjcf) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant_view.xml]: Readme folder was added.

File: purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py
Comment (javierjcf) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: I revert the cahnge.

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (pedrobaeza) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: It's needed, yes, because you will have a singleton error if not when having existing lines. It's a curious side effect here in v11.

File: purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py
Comment (javierjcf) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: I did this change again. because now, the wizard line has not field product_id, so the filtered was raising an error.

Comment (pedrobaeza) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: I have written about the same, but you are right. The product in current 2D matrix widget can be stored, so this is the less worst solution.

Comment (pedrobaeza) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: @simahawk this is a flaw in current matrix widget: you can't store a full record in the matrix. Only the fields in row, column and value are stored.

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (javierjcf) [inline purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: The product_id field was removed, so in the test file we can't do:
`  wizard.variant_line_ids.filtered(lambda x: x.product_id == product1).product_uom_qty = 0 `

Comment (pedrobaeza) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: Yeah, I saw it later on your other comment. Let's continue the conversation there.

**Comment by javierjcf** [comment]
> This was done on purpose for not being slave of the insertion order. Why do you remove it?

Because the product_id field was removed from _purchase.manage.variant.line_, so in the test file we can't do:
wizard.variant_line_ids.filtered(lambda x: x.product_id == product1).product_uom_qty = 0

File: purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py
Comment (simahawk) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: wild guess: can't we bypass this by making these fields invisible?
Otherwise: can you open an issue on OCA/web?

Comment (pedrobaeza) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: I'm afraid it doesn't serve. We were discussing in this PR: https://github.com/OCA/web/pull/1022, although it was more an internal chat between my colleague @Yajo and I.

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (javierjcf) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: Ahhh, OK i underestand the empty browse now, but with self.env['purchase.order.line'].create() we avoid any singleton error because we don't need to acces to purchase_order.order_line instance to create. I think that using self.env[] is the normal way to create a record.

Comment (pedrobaeza) [reply purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: Yeah, that's another option indeed. Here we make the browse for not having to enumerate the exact model, as we have it in the field, but if you see it more readable, please change it.

File: purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py
Comment (simahawk) [reply purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:106]: FTR reported here https://github.com/OCA/web/issues/1084

File: purchase_order_variant_mgmt/__manifest__.py
Comment (simahawk) [inline purchase_order_variant_mgmt/__manifest__.py:13]: https://github.com/OCA/product-variant

File: purchase_order_variant_mgmt/__manifest__.py
Comment (simahawk) [inline purchase_order_variant_mgmt/__manifest__.py]: remove demo

File: purchase_order_variant_mgmt/README.rst
Comment (simahawk) [inline purchase_order_variant_mgmt/README.rst]: for the next reviewer: this will be replaced by README generator ;)

File: purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py
Comment (simahawk) [inline purchase_order_variant_mgmt/tests/test_purchase_order_variant_mgmt.py:11]: disable tracking on the whole suite:
```py
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (simahawk) [inline purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: odoo

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (simahawk) [inline purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: would be nice to have params on new lines

File: purchase_order_variant_mgmt/wizard/purchase_manage_variant.py
Comment (simahawk) [inline purchase_order_variant_mgmt/wizard/purchase_manage_variant.py]: wouldn't be better to collect line values as `(0, 0, vals)` and write on the order once?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: purchase_order_variant_mgmt/README.rst
Comment (pedrobaeza) [reply purchase_order_variant_mgmt/README.rst]: I prefer to pre-generate the README on the migration PR, as you reduce diff size and avoid these confusions. @javierjcf maybe you can try to install maintainer-tools (with python3) and run oca-gen-addon-readme for having it.

**Comment by javierjcf** [comment]
I don't understand why now ist's raising a flake8 error in the travis integretion, in the product_variant_sale_price module, which is not changed in this PR.

this is the error:
`product_variant_sale_price/hooks.py:6: [E8103(sql-injection), set_sale_price_on_variant] SQL injection risk. Use parameters if you can. - More info https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection`

**Comment by pedrobaeza** [comment]
That warning was before, but it's not the one that provokes the red state. This is the error:
```
purchase_order_variant_mgmt/wizard/purchase_manage_variant.py:16:53: W291 trailing whitespace
```

**Comment by javierjcf** [comment]
> That warning was before, but it's not the one that provokes the red state. This is the error:
> 
> ```
> purchase_order_variant_mgmt/wizard/purchase_manage_variant.py:16:53: W291 trailing whitespace
> ```

Thanks @pedrobaeza. It's fixed now.

**Comment by GSLabIt** [comment]
is this ready to merge?

**Comment by pedrobaeza** [comment]
There are no approvals on the PR, so it should be reviewed and approved before being merged. @GSLabIt you can be one of them.

**Comment by GSLabIt** [comment]
@pedrobaeza will have a look in a few, thx

**Review by GSLabIt — APPROVED** [review]
Functional tests LGTM

**Comment by pedrobaeza** [comment]
@javierjcf please squash your migration commits together for merging this.

**Review by grindtildeath — COMMENTED** [review]
Tested and looks ok, however, the widget is limited for products having more than 2 variants, shouldn't we state this clearly in readme/ROADMAP.rst and maybe restrict the selection of product template to products having max 2 variants ?

File: purchase_order_variant_mgmt/__manifest__.py
Comment (grindtildeath) [reply purchase_order_variant_mgmt/__manifest__.py:13]: Still to attend

**Comment by pedrobaeza** [comment]
Superseeded by #152

**Comment by pedrobaeza** [comment]
Not really superseeded, so reopening.

**Comment by Roodin** [comment]
LGTM

**Comment by pedrobaeza** [comment]
Let's merge it as an old one and already migrated to upper versions: /ocabot merge

**Pull request merged**
